<a href="https://colab.research.google.com/github/alpha8108/mulcamp34/blob/main/pandas_240104_Copy1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
!pip install pandas


[notice] A new release of pip is available: 23.2.1 -> 23.3.2
[notice] To update, run: python.exe -m pip install --upgrade pip


In [ ]:
import pandas as pd
pd.__version__

'2.1.4'

In [ ]:
!pip install seaborn

p.237
## pandas의 데이터 구조
- Series 데이터 구조, 1차원 데이터 : 컬럼의 개수가 1개인 데이터
- DateFrame 데이터 구조, 컬럼 개수가 여러개인 데이터

In [ ]:
s1 = pd.Series([10, 20, 30, 40, 50])
s1

0    10
1    20
2    30
3    40
4    50
dtype: int64

In [ ]:
s1.index

RangeIndex(start=0, stop=5, step=1)

In [ ]:
s1.values

array([10, 20, 30, 40, 50], dtype=int64)

위에 벨류 저거 numpy배열에서 나왔잖아 numpy를 복습하쟈 ㅎ (기본을 탄탄히)

In [ ]:
s1 = pd.Series(['a', 'b', 'c', 1, 2, 3])
s1

0    a
1    b
2    c
3    1
4    2
5    3
dtype: object

문자와 숫자가 같이있다고해서 에러가나는 건 아님. object가뭐냐
문자로 처리하더라 이거임 숫자고 그냥 문자로 처리하는 것 일반적으로
엄밀히말하면 문자는 아님 object라는 타입이 있는데 편하게 생각하면 그렇다.

In [ ]:
import numpy as np

s3 = pd.Series([np.nan, 10, 30]) # nan= 결측치(값이 비어있따), missing values
s3

0     NaN
1    10.0
2    30.0
dtype: float64

In [ ]:
index_date = ['2018-10-07', '2018-10-08'] #인덱스 지정
s4 = pd.Series([200, 195], index = index_date)
s4

# index 설정의 기본 원칙은 중복이 되면 안된다. 코드 상으로는 중복은 허용하더라.

2018-10-07    200
2018-10-08    195
dtype: int64

In [ ]:
data_dict = {
    '국어' : 100,
    '영어' : 95
}

s5 = pd.Series(data_dict)
s5 #데이터 생성한거임 불러오는 것은 아직 하지 않았다.

국어    100
영어     95
dtype: int64

p.242
## 날짜 데이터


In [ ]:
pd.date_range(start='2024/01/01', end='2024.01.07') #p.242 (예제 참조)

DatetimeIndex(['2024-01-01', '2024-01-02', '2024-01-03', '2024-01-04',
               '2024-01-05', '2024-01-06', '2024-01-07'],
              dtype='datetime64[ns]', freq='D')

In [ ]:
pd.date_range(start='2024/01/01', periods = 7) #end 쓰기도 귀찮아

DatetimeIndex(['2024-01-01', '2024-01-02', '2024-01-03', '2024-01-04',
               '2024-01-05', '2024-01-06', '2024-01-07'],
              dtype='datetime64[ns]', freq='D')

판다스는 메서드수가 기하급수적으로 늘어남. 변수도 마찬가지. 내가 만약에
가까운 개념들이 궁금하다 싶으신 분들은
pandas.date_range들어가서 메서드 설명 보라. https://pandas.pydata.org/docs/reference/api/pandas.date_range.html
판다스는 메서드 보는거 능력 키워야함 특히
각각에 해당되는 것들이 반복문써서 지지고 볶고하는게
사실 이미 메서드에 녹여져있는 케이스도 많다.
파라미터(매개변수) 이런것도 잘 읽어서 공부해서
이거 이용해서 문제해결하려고 코드 익히는게 큰 도움이 됨.

In [ ]:
pd.date_range(start='2024/01/01', periods = 4, freq = '2D') #날짜 텀 지정

DatetimeIndex(['2024-01-01', '2024-01-03', '2024-01-05', '2024-01-07'], dtype='datetime64[ns]', freq='2D')

#시간대도 지정 가능 #교재보면 비슷한 예제인데 매개변수에대한 의미
#그리고 옵션을 어떻게 할 것인지 freq 이것만 알아도 외우려고하지말고
#어딘가에 참조할만한 링크만 걸어놔

pd.date_range(start='2024-01-01 08:00', periods = 4, freq = 'H')

## DateFrame을 활용한 데이터 생성

In [ ]:
import seaborn as sns # 시각화 라이브러리 (통계 시각화)

sns.__version__

'0.13.1'

In [ ]:
sns.get_dataset_names() #데이터 셋이라 생각행

['anagrams',
 'anscombe',
 'attention',
 'brain_networks',
 'car_crashes',
 'diamonds',
 'dots',
 'dowjones',
 'exercise',
 'flights',
 'fmri',
 'geyser',
 'glue',
 'healthexp',
 'iris',
 'mpg',
 'penguins',
 'planets',
 'seaice',
 'taxis',
 'tips',
 'titanic']

In [ ]:
#데이터 프레임 불러올껀디 잘봐 #판다스 샘플데이터는 이렇게 불러오면 된다.
iris = sns.load_dataset('iris')
iris

sepal_length  sepal_width  petal_length  petal_width    species
0             5.1          3.5           1.4          0.2     setosa
1             4.9          3.0           1.4          0.2     setosa
2             4.7          3.2           1.3          0.2     setosa
3             4.6          3.1           1.5          0.2     setosa
4             5.0          3.6           1.4          0.2     setosa
..            ...          ...           ...          ...        ...
145           6.7          3.0           5.2          2.3  virginica
146           6.3          2.5           5.0          1.9  virginica
147           6.5          3.0           5.2          2.0  virginica
148           6.2          3.4           5.4          2.3  virginica
149           5.9          3.0           5.1          1.8  virginica

[150 rows x 5 columns]

In [ ]:
# p247 여기는 참고만
# p249 와강튼 딕셔너리 스타일을 좋아함 (그래도 다른거 타이핑 해보고 넘어가셈)
# 입문 시 괄호 자동완성은 일단 비추 실무에서부터 자동완성 추천

table_data = {
    '연도' : [2015, 2016, 2016, 2017, 2017], #키값이 컬럼으로, 리스트가 각각의 데이터값이된것
    '지사' : ['한국', '한국', '미국', '한국', '미국'],
    '고객 수' : [200, 250, 450, 300, 500]
}

table_data

{'연도': [2015, 2016, 2016, 2017, 2017],
 '지사': ['한국', '한국', '미국', '한국', '미국'],
 '고객 수': [200, 250, 450, 300, 500]}

In [ ]:
pd.DataFrame(table_data) # 강사가 선호하는 방법 / 딕셔너리형태

연도  지사  고객 수
0  2015  한국   200
1  2016  한국   250
2  2016  미국   450
3  2017  한국   300
4  2017  미국   500

In [ ]:
data = pd.DataFrame(table_data) #데이터로 저장
data

연도  지사  고객 수
0  2015  한국   200
1  2016  한국   250
2  2016  미국   450
3  2017  한국   300
4  2017  미국   500

In [ ]:
data.index

RangeIndex(start=0, stop=5, step=1)

In [ ]:
data.values #몇차원인지 헷갈리면 .shape붙여라 얘는 2차원

array([[2015, '한국', 200],
       [2016, '한국', 250],
       [2016, '미국', 450],
       [2017, '한국', 300],
       [2017, '미국', 500]], dtype=object)

실무에서는
- 데이터 가공 시, numpy 메서드와 pandas 메서드 조합을 해서 처리하는 경우 많음
  + vectorization 라고하는데 / 파이썬 기초문법 (for-loop같은거)대신
  + 속도가 매우 빠름 / vectorization이 궁금하다 그럼gpt한테 빨리 물어봐. 키워드알게되면
  + 이런거 알게 될때마다 블로그같은거에 기록해놓으삼 회사가면 좋아함 걔네도 잘 못해 ㅋ

In [ ]:
data.columns

Index(['연도', '지사', '고객 수'], dtype='object')

In [ ]:
titanic = sns.load_dataset('titanic') # 4교시 셋팅
titanic

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

## 데이터 연산

In [ ]:
!pip install seaborn


[notice] A new release of pip is available: 23.3.1 -> 23.3.2
[notice] To update, run: python.exe -m pip install --upgrade pip


In [ ]:
#넘파이와 유사하다는걸 느낄 것이다.
# No module named 'seaborn' --> 라이브러리 설치 안됐다는 뜻

import pandas as pd
import seaborn as sns
import numpy as np


s1 = pd.Series([1, 2, 3, 4])
s2 = pd.Series([10, 20, 30, 40, 50])
s1 + s2

0    11.0
1    22.0
2    33.0
3    44.0
4     NaN
dtype: float64

In [ ]:
s2 / s1

0    10.0
1    10.0
2    10.0
3    10.0
4     NaN
dtype: float64

In [ ]:
s1 * s2

0     10.0
1     40.0
2     90.0
3    160.0
4      NaN
dtype: float64

In [ ]:
# 리스트와 Numpy의 배열과 다르게, 서로 크기가 달라도 연산할 수 있음

s1 = pd.Series([1, 2, 3, 4])
s2 = pd.Series([10, 20, 30, 40, 50])
s1 + s2  #NaN = 결측치

0    11.0
1    22.0
2    33.0
3    44.0
4     NaN
dtype: float64

In [ ]:
table_data1 = {'A': [1, 2, 3, 4, 5],
              'B': [10, 20, 30, 40, 50],
              'C': [100, 200, 300, 400, 500]}
df1 = pd.DataFrame(table_data1)
df1 #슬랙복붙

A   B    C
0  1  10  100
1  2  20  200
2  3  30  300
3  4  40  400
4  5  50  500

In [ ]:
table_data2 = {'A': [6, 7, 8],
              'B': [60, 70, 80],
              'C': [600, 700, 800]}
df2 = pd.DataFrame(table_data2)
df2 #슬랙복붙  #행의크기는 다르지만 열의 크기가 같으니 연산할 수 있다.

A   B    C
0  6  60  600
1  7  70  700
2  8  80  800

In [ ]:
df1 + df2 # 안맞는 애들은 NaN이 나온다. 결측치로 빠지게 됨 연산이 안됐으니까.

A      B       C
0   7.0   70.0   700.0
1   9.0   90.0   900.0
2  11.0  110.0  1100.0
3   NaN    NaN     NaN
4   NaN    NaN     NaN

In [ ]:
#사칙연산 되는 지 보자
df1 * df2

A       B         C
0   6.0   600.0   60000.0
1  14.0  1400.0  140000.0
2  24.0  2400.0  240000.0
3   NaN     NaN       NaN
4   NaN     NaN       NaN

In [ ]:
df1 /df2 # 잘되는구나

A         B         C
0  0.166667  0.166667  0.166667
1  0.285714  0.285714  0.285714
2  0.375000  0.375000  0.375000
3       NaN       NaN       NaN
4       NaN       NaN       NaN

In [ ]:
table_data3 = {'봄':  [256.5, 264.3, 215.9, 223.2, 312.8],
              '여름': [770.6, 567.5, 599.8, 387.1, 446.2],
              '가을': [363.5, 231.2, 293.1, 247.7, 381.6],
              '겨울': [139.3, 59.9, 76.9, 109.1, 108.1]}
columns_list = ['봄', '여름', '가을', '겨울']
index_list = ['2012', '2013', '2014', '2015', '2016']

df3 = pd.DataFrame(table_data3, columns = columns_list, index = index_list) #이 줄은 복붙하면 안된다.
df3 #복붙할 때 주의 할점 바로 윗줄복붙x 매개명수가 생각이 안난다. 또 구글링 하게된다. 생산성 하락. 흐름을 이해하면서 타이핑
# 컬럼명 바꾸고 싶으면 columns 를 원하는 이름으로 바꿔주면 된다.
#그리고 여기서 컬럼리스트 굳이 안써도 상관은 없다고 하나 그건 위(table)에 이미 키값이 정리가 되어있기 떄문이다.
# 만약 그냥 인덱스 번호가 뜰 수 있잖아 그럼 컬럼명을 변경해주는게 좋겠지?

봄     여름     가을     겨울
2012  256.5  770.6  363.5  139.3
2013  264.3  567.5  231.2   59.9
2014  215.9  599.8  293.1   76.9
2015  223.2  387.1  247.7  109.1
2016  312.8  446.2  381.6  108.1

In [ ]:
#평균이랑 표준편차를 구해보자

df3.mean() #칼럼별로 들어가게 됨 각 칼럼에 대한 평균값

봄     254.54
여름    554.24
가을    303.42
겨울     98.66
dtype: float64

In [ ]:
df3.max() # 각 컬럼의 최댓값이 나왔땅

봄     312.8
여름    770.6
가을    381.6
겨울    139.3
dtype: float64

In [ ]:
df3.min() #최소값

봄     215.9
여름    387.1
가을    231.2
겨울     59.9
dtype: float64

In [ ]:
df3.std() # 표준편차 # 엑셀 생각해봐라 판다스가 더 편하다.

봄      38.628267
여름    148.888895
가을     67.358496
겨울     30.925523
dtype: float64

In [ ]:
df3.mean(axis=1) #p256 # 자 이게 어떤 의미인가. --> 행의 평균값. axis 이건 행단위로 직계연산을 한다는 뜻

2012    382.475
2013    280.725
2014    296.425
2015    241.775
2016    312.175
dtype: float64

In [ ]:
df3 #이거 보면서 위에 저게 뭔지 생각해봐.

봄     여름     가을     겨울
2012  256.5  770.6  363.5  139.3
2013  264.3  567.5  231.2   59.9
2014  215.9  599.8  293.1   76.9
2015  223.2  387.1  247.7  109.1
2016  312.8  446.2  381.6  108.1

In [ ]:
df3.max(axis=1) #이해 되는가? #이거 강수량데이터라 여름만 나온거 ㅋ 최대값

2012    770.6
2013    567.5
2014    599.8
2015    387.1
2016    446.2
dtype: float64

In [ ]:
df3.std(axis=1) # 행단위로 표준편차를 구했다는 것 #상상하는 능력을 키우는게 핵심.

2012    274.472128
2013    211.128782
2014    221.150739
2015    114.166760
2016    146.548658
dtype: float64

  밑은 평균 위는 표준편차 .   2012년예를들면
  평균 382부터 편차가 앞뒤로 274정도 와리가리 했다는 것이다.
  2025년은 그럼 241부터 앞뒤로 114정도씩 와리가리 했다는 것이다.

In [ ]:
df3.mean(axis=1) #해석하는 능력을 키우는게 분석! 데이터를 가지고 상상해 분포가 대충 어떤 그림일것이다.

2012    382.475
2013    280.725
2014    296.425
2015    241.775
2016    312.175
dtype: float64

In [ ]:
df3.describe() # 사실 위의 메서드를 하나하나 보는게 아니라 이렇게 보고 상상해 데이터분포를.

봄          여름          가을          겨울
count    5.000000    5.000000    5.000000    5.000000
mean   254.540000  554.240000  303.420000   98.660000
std     38.628267  148.888895   67.358496   30.925523
min    215.900000  387.100000  231.200000   59.900000
25%    223.200000  446.200000  247.700000   76.900000
50%    256.500000  567.500000  293.100000  108.100000
75%    264.300000  599.800000  363.500000  109.100000
max    312.800000  770.600000  381.600000  139.300000

In [ ]:
import pandas as pd
import numpy as np

KTX_data = {'경부선 KTX': [39060, 39896, 42005, 43621, 41702, 41266, 32427],
            '호남선 KTX': [7313, 6967, 6873, 6626, 8675, 10622, 9228],
            '경전선 KTX': [3627, 4168, 4088, 4424, 4606, 4984, 5570],
            '전라선 KTX': [309, 1771, 1954, 2244, 3146, 3945, 5766],
            '동해선 KTX': [np.nan,np.nan, np.nan, np.nan, 2395, 3786, 6667]}
index_list = ['2011', '2012', '2013', '2014', '2015', '2016', '2017']

df_KTX = pd.DataFrame(KTX_data, index = index_list)
df_KTX

경부선 KTX  호남선 KTX  경전선 KTX  전라선 KTX  동해선 KTX
2011    39060     7313     3627      309      NaN
2012    39896     6967     4168     1771      NaN
2013    42005     6873     4088     1954      NaN
2014    43621     6626     4424     2244      NaN
2015    41702     8675     4606     3146   2395.0
2016    41266    10622     4984     3945   3786.0
2017    32427     9228     5570     5766   6667.0

In [ ]:
#p.259 https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html
#왼쪽에 해당되는 메서드들이 존재하는데
#찾아서 쓰라. 오늘은 head 써봄
#여기 보면서 문장보고 샘플코드 보면서
#장고수업 보기 전까지 훈련을 해라.
df_KTX.head() #상단5개 출력이 되는디.

경부선 KTX  호남선 KTX  경전선 KTX  전라선 KTX  동해선 KTX
2011    39060     7313     3627      309      NaN
2012    39896     6967     4168     1771      NaN
2013    42005     6873     4088     1954      NaN
2014    43621     6626     4424     2244      NaN
2015    41702     8675     4606     3146   2395.0

In [ ]:
df_KTX.head(3) # 처음부터 3개반환

경부선 KTX  호남선 KTX  경전선 KTX  전라선 KTX  동해선 KTX
2011    39060     7313     3627      309      NaN
2012    39896     6967     4168     1771      NaN
2013    42005     6873     4088     1954      NaN

In [ ]:
df_KTX.tail(3) #끝에서부터 3개반환한다는 뜻


경부선 KTX  호남선 KTX  경전선 KTX  전라선 KTX  동해선 KTX
2015    41702     8675     4606     3146   2395.0
2016    41266    10622     4984     3945   3786.0
2017    32427     9228     5570     5766   6667.0

## 행 선택(=행 추출)
- slicing과 동일한 문법으로 추출 가능

In [ ]:
df_KTX[0:2]

경부선 KTX  호남선 KTX  경전선 KTX  전라선 KTX  동해선 KTX
2011    39060     7313     3627      309      NaN
2012    39896     6967     4168     1771      NaN

In [ ]:
df_KTX[2:5]

경부선 KTX  호남선 KTX  경전선 KTX  전라선 KTX  동해선 KTX
2013    42005     6873     4088     1954      NaN
2014    43621     6626     4424     2244      NaN
2015    41702     8675     4606     3146   2395.0

## loc vs iloc를 활용한 열 추출, 행 추출
- 두가지의 차이 기억하기
- 한가지만 기억하고자 한다면, loc만 기억하자. (오늘의 블로그 주제)
- 코드작성할 때 가급적이면 loc로 하자 iloc말고

In [ ]:
#df_KTX.loc[행, 열] #이게 기준
#df_KTX.loc['2013':'2016'] #좋은 문법은 아니다.

경부선 KTX  호남선 KTX  경전선 KTX  전라선 KTX  동해선 KTX
2013    42005     6873     4088     1954      NaN
2014    43621     6626     4424     2244      NaN
2015    41702     8675     4606     3146   2395.0
2016    41266    10622     4984     3945   3786.0

In [ ]:
df_KTX.loc['2013':'2016', '경부선 KTX'] #이 코드를 기억하면 웬만한 행추출 쌉가능 다른 코드들은 지저분

2013    42005
2014    43621
2015    41702
2016    41266
Name: 경부선 KTX, dtype: int64

In [ ]:
#P.264는 그냥 통으로 삭제시켜 너무 지저분해

In [ ]:
df_KTX.loc['2013':'2016', ['경부선 KTX','호남선 KTX']]

경부선 KTX  호남선 KTX
2013    42005     6873
2014    43621     6626
2015    41702     8675
2016    41266    10622

In [ ]:
df_KTX.loc['2013':'2016', ['호남선 KTX']] #나오긴함

호남선 KTX
2013     6873
2014     6626
2015     8675
2016    10622

In [ ]:
iris['sepal_length'] >=7

0      False
1      False
2      False
3      False
4      False
       ...  
145    False
146    False
147    False
148    False
149    False
Name: sepal_length, Length: 150, dtype: bool

In [ ]:
# df_KTX.loc[행조건식, [컬럼명]]  이걸 디폴트로 외우면 된다.

iris = sns.load_dataset('iris')

# sepal_length의 길이가 5cm 이상인 것만 조회
iris.loc[iris['sepal_length'] >=7 , :] #[:, :] 모든데이터 가져온단 뜻 자 그럼
# 그럼 행조건식 넣은게 True/False 나올거잖음 iris['sepal_length'] >=7 이거 --> 트루값만 가져온거지 그럼?
# 그니까 조건식은 True/False 나오게 적는게 포인트라는 것 상기

sepal_length  sepal_width  petal_length  petal_width     species
50            7.0          3.2           4.7          1.4  versicolor
102           7.1          3.0           5.9          2.1   virginica
105           7.6          3.0           6.6          2.1   virginica
107           7.3          2.9           6.3          1.8   virginica
109           7.2          3.6           6.1          2.5   virginica
117           7.7          3.8           6.7          2.2   virginica
118           7.7          2.6           6.9          2.3   virginica
122           7.7          2.8           6.7          2.0   virginica
125           7.2          3.2           6.0          1.8   virginica
129           7.2          3.0           5.8          1.6   virginica
130           7.4          2.8           6.1          1.9   virginica
131           7.9          3.8           6.4          2.0   virginica
135           7.7          3.0           6.1          2.3   virginica

In [ ]:
# species가 virginica인 것만 조회
iris.loc[iris['virginica'],['species']] # 걍 모르겠어욥 ㅎ....
#--> iris.loc[iris['컬럼명'] == 'virginica',:]] (피드백)

#case 1
# iris.loc[iris['species'] == 'virginica', :] 행조건만 맞추면 된다! True/False나오게 ! 문자끼리 할 때는 == 두개!
그리고



# df_KTX.loc[행조건식, [컬럼명]]  이걸 디폴트로 외우면 된다.

KeyError: 'virginica'

In [ ]:
#문제2
titanic = sns.load_dataset('titanic')
titanic.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [ ]:
# embark_town 컬럼에서 Southampton 거주민의 모든 데이터를 조회하세요
# iris.loc[iris['컬럼명'] == 'virginica',:]] # : 모든 컬럼 조회 처음에는 다 조회하고 나중에는 나만의 스타일로

titanic.loc[titanic['embark_town'] == 'Southampton',:] #대소문자 구분 잘하셈
# 팁 테이블이나 데이터에서 복사해서 넣을 수 있으면 복사해서 넣어라 저거 Southampton 같은거



survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
6           0       1    male  54.0      0      0  51.8625        S   First   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
883         0       2    male  28.0      0      0  10.5000        S  Second   
884         0       3    male  25.0      0      0   7.0500        S   Third   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
6      man        True    E  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
883    man        True  NaN  Southampton    no   True  
884    man        True  NaN  Southampton    no   True  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  

[644 rows x 15 columns]

In [ ]:
# age 평균 구하고, 평균 이상인 데이터만 조회하기
# titanic.loc[:,'age'].mean() 난 평균을 어렵게 구했네...
titanic.loc[titanic['age'] >= 29.69911764705882, :]

# 강사의 평균 구하는 법
# titanic['age'].mean()
# titanic.loc[titanic['age'] >= titanic['age'].mean(), :].

# 그다음에
# var = 'age'
# 한다음에 'age' 자리에 var입력


survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
1           1       1  female  38.0      1      0  71.2833        C  First   
3           1       1  female  35.0      1      0  53.1000        S  First   
4           0       3    male  35.0      0      0   8.0500        S  Third   
6           0       1    male  54.0      0      0  51.8625        S  First   
11          1       1  female  58.0      0      0  26.5500        S  First   
..        ...     ...     ...   ...    ...    ...      ...      ...    ...   
873         0       3    male  47.0      0      0   9.0000        S  Third   
879         1       1  female  56.0      0      1  83.1583        C  First   
881         0       3    male  33.0      0      0   7.8958        S  Third   
885         0       3  female  39.0      0      5  29.1250        Q  Third   
890         0       3    male  32.0      0      0   7.7500        Q  Third   

       who  adult_male deck  embark_town alive  alone  
1    woman       False    C    Cherbourg   yes  False  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
6      man        True    E  Southampton    no   True  
11   woman       False    C  Southampton   yes   True  
..     ...         ...  ...          ...   ...    ...  
873    man        True  NaN  Southampton    no   True  
879  woman       False    C    Cherbourg   yes  False  
881    man        True  NaN  Southampton    no   True  
885  woman       False  NaN   Queenstown    no  False  
890    man        True  NaN   Queenstown    no   True  

[330 rows x 15 columns]

In [ ]:
#문제 하나 더
tips = sns.load_dataset('tips')
tips.head()

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

In [ ]:
# total_bill의 중간값 이하인 것만 조회 (복습ㄱㄱ)
# tips['total_bill'].median()
tips.loc[tips['total_bill'] <= tips['total_bill'].median(), : ].head(1) # head는 그냥 위에서 뽑고싶어서 덧붙힘

# day가 Sun인 데이터만 조회하기 (복습ㄱㄱ)

# tip의 평균보다 크면서, time이 Dinner인 데이터만 조회


total_bill   tip     sex smoker  day    time  size
0         16.99  1.01  Female     No  Sun  Dinner     2
1         10.34  1.66    Male     No  Sun  Dinner     3
6          8.77  2.00    Male     No  Sun  Dinner     2
8         15.04  1.96    Male     No  Sun  Dinner     2
9         14.78  3.23    Male     No  Sun  Dinner     2
..          ...   ...     ...    ...  ...     ...   ...
232       11.61  3.39    Male     No  Sat  Dinner     2
233       10.77  1.47    Male     No  Sat  Dinner     2
234       15.53  3.00    Male    Yes  Sat  Dinner     2
235       10.07  1.25    Male     No  Sat  Dinner     2
236       12.60  1.00    Male    Yes  Sat  Dinner     2

[122 rows x 7 columns]

In [ ]:
# day가 Sun인 데이터만 조회하기 (복습ㄱㄱ)
tips.loc[tips['day']=='Sun',:]

total_bill   tip     sex smoker  day    time  size
0         16.99  1.01  Female     No  Sun  Dinner     2
1         10.34  1.66    Male     No  Sun  Dinner     3
2         21.01  3.50    Male     No  Sun  Dinner     3
3         23.68  3.31    Male     No  Sun  Dinner     2
4         24.59  3.61  Female     No  Sun  Dinner     4
..          ...   ...     ...    ...  ...     ...   ...
186       20.90  3.50  Female    Yes  Sun  Dinner     3
187       30.46  2.00    Male    Yes  Sun  Dinner     5
188       18.15  3.50  Female    Yes  Sun  Dinner     3
189       23.10  4.00    Male    Yes  Sun  Dinner     3
190       15.69  1.50    Male    Yes  Sun  Dinner     2

[76 rows x 7 columns]

In [ ]:
# tip의 평균보다 크면서, time이 Dinner인 데이터만 조회
# titanic['age'].mean()
#셋팅을 먼저 해 tips.loc[(조건1) & (조건2),:] ---> 여기서 시작해 (다중조건)

#tips.loc[(조건1) & (조건2),:] #다중조건을 무한으로 쓸 수 있는 것임 (tips['size']==3) & (tips['sex'] == 'Male')
# 이런것들을 계속 더 넣어도 됨


tips.loc[(tips['tip'] > tips['tip'].mean()) & (tips['time'] == 'Dinner'), :]

#한번에 두개 쓰기 힘드니까 하나씩 써서 실행해보고 복붙해서 넣어도 됨.

#코드 길어지니까 :] 이거 미리 내려놓으삼

#and 조건, a이고 b이다. a이면서b이다.
#or 조건, a이거나 또는 b이다.
# &, AND조건 / \(엔터위에), OR조건
#조건식 두개 써야겠죠?

total_bill   tip     sex smoker   day    time  size
2         21.01  3.50    Male     No   Sun  Dinner     3
3         23.68  3.31    Male     No   Sun  Dinner     2
4         24.59  3.61  Female     No   Sun  Dinner     4
5         25.29  4.71    Male     No   Sun  Dinner     4
7         26.88  3.12    Male     No   Sun  Dinner     4
..          ...   ...     ...    ...   ...     ...   ...
232       11.61  3.39    Male     No   Sat  Dinner     2
234       15.53  3.00    Male    Yes   Sat  Dinner     2
238       35.83  4.67  Female     No   Sat  Dinner     3
239       29.03  5.92    Male     No   Sat  Dinner     3
243       18.78  3.00  Female     No  Thur  Dinner     2

[98 rows x 7 columns]

In [ ]:
#위에 다중조건 예시
tips.loc[(tips['tip'] >= tips['tip'].mean()) &
    (tips['time'] == 'Dinner') &
    (tips['size'] == 3) &
    (tips['sex'] == 'Male') &
    (tips['day'] == 'Sat'),
:]

total_bill    tip   sex smoker  day    time  size
19        20.65   3.35  Male     No  Sat  Dinner     3
35        24.06   3.60  Male     No  Sat  Dinner     3
39        31.27   5.00  Male     No  Sat  Dinner     3
65        20.08   3.15  Male     No  Sat  Dinner     3
170       50.81  10.00  Male    Yes  Sat  Dinner     3
206       26.59   3.41  Male    Yes  Sat  Dinner     3
231       15.69   3.00  Male    Yes  Sat  Dinner     3
239       29.03   5.92  Male     No  Sat  Dinner     3